### PR TITLE
feat(node): worker_threads

### PR DESCRIPTION
**Pull request by Mesteery**

### CODE REVIEW

**Comment by Mesteery** [comment]
CI errors are related to https://github.com/denoland/deno/issues/9696.

**Review by bartlomieju — CHANGES_REQUESTED** [review]
This looks good @Mesteery, but please some more tests that actually execute some code in workers.

**Comment by Mesteery** [comment]
Yes of course, the more complete tests are already added locally.
The problem is that it's not completely ready, since the implementation of `events` differs from Node.js (and this causes some problems with `worker_threads`). So I will rewrite it.

File: node/worker_threads.ts [hunk 1]
```diff
@@ -0,0 +1,217 @@
+/// <reference lib="webworker"/>
```
[the remark this answers is not in the document]
Comment (bartlomieju) [reply node/worker_threads.ts:1]: @kitsonk could you please advice on this situation?

**Comment by bartlomieju** [comment]
Ooops, looks like I borked the merge.

**Comment by Mesteery** [comment]
no worries, it's fixed

**Comment by guybedford** [comment]
It's a shame this one got derailed, the PR seemed to be very close!

**Comment by bartlomieju** [comment]
@guybedford this PR is more or less complete, however there's still problem with TS types. I wonder if we should maybe skip typings completely for the time being and just land it.

**Comment by kt3k** [comment]
I think I managed to fix type errors. PTAL @bartlomieju @kitsonk

**Comment by bartlomieju** [comment]
@kt3k this seems fine to me, but I will let @kitsonk take a look before landing.

**Comment by bartlomieju** [comment]
This is a blocker for `parcel`, so I'm going to land this PR tonight.

**Review by littledivy — CHANGES_REQUESTED** [review]
`require` and compat APIs are not available to workers. Can we please add that?

**Comment by Mesteery** [comment]
> require and compat APIs are not available to workers. Can we please add that? EDIT: Needs support in CLI

Sorry but I am not familiar with Deno. What do you mean by that?

**Comment by bartlomieju** [comment]
> > require and compat APIs are not available to workers. Can we please add that? EDIT: Needs support in CLI
> 
> Sorry but I am not familiar with Deno. What do you mean by that?

This is something I will work on, currently worker API doesn't respect "compat" mode and doesn't add required globals.

File: node/worker_threads.ts [hunk 2]
```diff
@@ -1,63 +1,234 @@
...
+  parentPort.eventNames = () => [""];
+  parentPort.listenerCount = () => 0;
+
+  parentPort.emit = () => notImplemented();
+  parentPort.removeAllListeners = () => notImplemented();
+
+  [{ threadId, workerData, environmentData }] = await once(
+    parentPort,
+    "message",
+  );
```
Comment (bartlomieju) [inline node/worker_threads.ts:179]: This line causes following error to be raised:
```
error: Uncaught (in worker "") Module evaluation is still pending but there are no pending ops or dynamic imports. This situation is often caused by unresolved promises.
error: Uncaught (in promise) Error: Unhandled error. ('Module evaluation is still pending but there are no pending ops or dynamic imports. This situation is often caused by unresolved promises.')
    const err = new ERR_UNHANDLED_ERROR(stringifiedEr);
```

**Comment by bartlomieju** [comment]
After a lot back and forth I found the root problem of the implementation of this PR. Currently tests fails with "Module execution is still pending but there are no pending ops..." suggesting that there's a promise that never resolves. I found that this is caused by `parentHandle` (which is somewhat an alias of `self`) is not actually an `EventEmitter` instance and so events are not properly dispatched. I'm not sure how to solve this problem, if I understood correctly we would have to make `self` (in the worker context) an instance of `EventEmitter` but I'm unsure how this could be achieved. Any ideas are welcome.

Alternatively we would have apply a lot of code from `EventEmitter` manually to copy the behavior (though it still might not be enough due to `instanceof` checks)

**Review by bartlomieju — APPROVED** [review]
LGTM, thanks for finishing this @kt3k. And thank you to @Mesteery sorry it took so long to land this!

**Review by littledivy — APPROVED** [review]
:shipit:

**Review by kt3k — APPROVED** [review]
@Mesteery Thank you for your initial work!

**Pull request merged**